src/compiler/parser.ts

```diff
@@ -6327,8 +6328,18 @@ namespace Parser {
         Debug.assert(languageVariant !== LanguageVariant.JSX, "Type assertions should never be parsed in JSX; they should be parsed as comparisons or JSX elements/fragments.");
         const pos = getNodePos();
         parseExpected(SyntaxKind.LessThanToken);
+        const posLessThan = getNodePos();
         const type = parseType();
-        parseExpected(SyntaxKind.GreaterThanToken);
+
+        // <Type />: specialized diagnostic for a JSX element in a .ts file - and skip some excess reporting
+        if (token() === SyntaxKind.SlashToken && scanner.scan() === SyntaxKind.GreaterThanToken) {
+            nextToken();
+            parseErrorAt(posLessThan - 1, getNodePos(), Diagnostics.JSX_tags_are_not_permitted_in_ts_files_Did_you_mean_to_change_the_file_extension_to_tsx, tokenToString(SyntaxKind.GreaterThanToken));
+        }

```

Comment:
I'd originally had it return here, so it doesn't try to `parseSimpleUnaryExpression`. But I couldn't figure out how to make an empty identifier that obeyed node ordering rules. See failures from up till [fd5adf7](https://github.com/microsoft/TypeScript/pull/56101/commits/fd5adf73cfd89c33c2e891f4519cf445aa774b78).

Reply:
Hard to say. I would personally assume rather low given that most JSX element children are often one of:

* More JSX
* Plain text: will often be parsed mostly as identifiers and/or odd punctuation characters
* `{}` blocks containing a function call or some other odd code - which is sometimes valid non-JSX